go/src/staleFlagCleaner.go

```diff
@@ -0,0 +1,624 @@
+				break
+			}
+		}
+	}
+
+	switch d := callExpr.Fun.(type) {
+	case *dst.Ident:
+		apiName = d.Name
+	case *dst.SelectorExpr:
+		apiName = d.Sel.Name
+	default:
+		// other cases will come in future
+	}
+
+	if element, ok := sfc.functype[apiName]; ok {
+		if element.FlagIndex == flagIndex {
+			switch element.OfType {
+			case treated:
+				return isTreated
+			case control:
+				return isControl
+			default:
+				return isUnknown
+			}
+		}
+	}
+	return isUnknown
+}
+
+/* This function will evaluate all expression nodes here */
+func (sfc *staleFlagCleaner) evaluateExprNode(exprNode dst.Expr) Value {
+
+	switch exprUnderConsideration := exprNode.(type) {
+	case *dst.Ident:
+		if val, ok := sfc.valueMap[exprUnderConsideration.Name]; ok {
+			return val
+		}
+		switch exprUnderConsideration.Name {
+		case strTrue:
+			return isTrue
+		case strFalse:
+			return isFalse
+		}
+
+	// Handling expressions related to calling a function
+	case *dst.CallExpr:
+		// Get the type of API
+		typeOfAPI := sfc.flagTypeAPI(exprUnderConsideration)
+		// Hardcoding the output of some API types
+		if typeOfAPI == isTreated {
+			if sfc.isTreated {
+				return isTrue
+			}
+			return isFalse
+		}
+		if typeOfAPI == isControl {
+			if sfc.isTreated {
+				return isFalse
+			}
+			return isTrue
+		}
+	//Binary expressions
+	case *dst.BinaryExpr:
+		valX := sfc.evaluateExprNode(exprUnderConsideration.X)
+		valY := sfc.evaluateExprNode(exprUnderConsideration.Y)
+		if valX != isBot || valY != isBot {
+			switch exprUnderConsideration.Op {
+			case token.LAND: //stands for &&
+				if valX == isFalse || valY == isFalse {
+					return isFalse
+				}
+				if valX == isTrue && valY == isTrue {
+					return isTrue
+				}
+			case token.LOR: //stands for ||
+				if valX == isTrue || valY == isTrue {
+					return isTrue
+				}
+				if valX == isFalse && valY == isFalse {
+					return isFalse
+				}
+			case token.EQL:
+				if valX != isBot && valY != isBot {
+					if valX == valY {
+						return isTrue
+					}
+					return isFalse
+				}
+
+			case token.NEQ:
+				if valX == valY {
+					return isFalse
+				}
+				return isTrue
+			}
+		}
+	//Unary expressions
+	case *dst.UnaryExpr:
+		switch exprUnderConsideration.Op {
+		case token.NOT:
+			valX := sfc.evaluateExprNode(exprUnderConsideration.X)
+			switch valX {
+			case isTrue:
+				return isFalse
+			case isFalse:
+				return isTrue
+			}
+		case token.AND: // &expr
+			return sfc.evaluateExprNode(exprUnderConsideration.X)
+		default:
+			//do nothing for now.
+		}
+	//Star expression (*expr)
+	case *dst.StarExpr:
+		return sfc.evaluateExprNode(exprUnderConsideration.X)
+	//Paren Expr: exprs inside parenthesis
+	case *dst.ParenExpr:
+		return sfc.evaluateExprNode(exprUnderConsideration.X)
+
+	//Selector Expr (expr.A, a.b, etc.)
+	case *dst.SelectorExpr:
+		switch ident := exprUnderConsideration.X.(type) {
+		case *dst.Ident:
+			if val, ok := sfc.valueMap[ident.Name+"."+exprUnderConsideration.Sel.Name]; ok {
```

Comment:
Is this correct? What if `ident` represents a local variable shadowing a global one, for example? Could the entry in `valueMap` be incorrect in this case, and contain entry for the global variable? Am I missing something?

Reply:
I think I haven't considered the case for the shadowing of the global variables. You are right.